examples/dtls/cid/server/main.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"sync/atomic"
+	"time"
+
+	piondtls "github.com/pion/dtls/v2"
+	"github.com/plgd-dev/go-coap/v3/dtls/server"
+	"github.com/plgd-dev/go-coap/v3/message"
+	"github.com/plgd-dev/go-coap/v3/message/codes"
+	"github.com/plgd-dev/go-coap/v3/mux"
+	"github.com/plgd-dev/go-coap/v3/options"
+	udpClient "github.com/plgd-dev/go-coap/v3/udp/client"
+)
+
+func handleA(w mux.ResponseWriter, r *mux.Message) {
+	log.Printf("got message in handleA:  %+v from %v\n", r, w.Conn().RemoteAddr())
+	err := w.SetResponse(codes.GET, message.TextPlain, bytes.NewReader([]byte("A hello world")))
+	if err != nil {
+		log.Printf("cannot set response: %v", err)
+	}
+}
+
+func handleB(w mux.ResponseWriter, r *mux.Message) {
+	log.Printf("got message in handleB:  %+v from %v\n", r, w.Conn().RemoteAddr())
+	customResp := w.Conn().AcquireMessage(r.Context())
+	defer w.Conn().ReleaseMessage(customResp)
+	customResp.SetCode(codes.Content)
+	customResp.SetToken(r.Token())
+	customResp.SetContentFormat(message.TextPlain)
+	customResp.SetBody(bytes.NewReader([]byte("B hello world")))
+	err := w.Conn().WriteMessage(customResp)
+	if err != nil {
+		log.Printf("cannot set response: %v", err)
+	}
+}
+
+// wrappedListener wraps a net.Listener and implements a go-coap DTLS
+// server.Listener.
+// NOTE: this utility is for example purposes only. Context should be handled
+// properly in meaningful scenarios.
+type wrappedListener struct {
+	l      net.Listener
+	closed atomic.Bool
+}
```

Comment:
@jkralik I'm not sure if there is appetite to do so, but I would be interested in `go-coap` defining a common `ContextListener` and utility wrappers to allow for the passing of `net.Listener`. For now, consumers of this functionality will need to use a custom listener as demonstrated in this example.

Reply:
> But one thing - Has handshaking been moved to Conn.Read/Write from listener.Accept? If not, we need to incorporate it with https://github.com/plgd-dev/go-coap/blob/master/net/dtlslistener.go#L37.

@jkralik I'm not sure I'm following your suggestion here. Handshaking has not moved -- from the perspective of the existing `DTLSListener` nothing has changed. In fact, setting the `connectionIDGenerator` on the `pion/dtls` config passed to `NewDTLSListener` would enable connection ID support. The issue is that the underlying `pion/transport/udp` package [being used](https://github.com/plgd-dev/go-coap/blob/12f9eb2affdaa01ec1a7c64cef4b3dad41f39241/net/dtlslistener.go#L95) does not support routing based on Connection ID. A replacement has been introduced in https://github.com/pion/dtls/tree/master/internal/net/udp, which is in `internal` for now so that it can be iterated upon prior to folks taking a direct dependency on its API. It can be consumed by using the [`pion/dtls.Listen` function](https://github.com/pion/dtls/blob/609e5bee6eb0269574fed9f5d50127fdb360b846/listener.go#L16).

I do think it would be good for the functionality to become the default in `go-coap`, but right now the `DTLSListener` is restrictive about the behavior it allows (anecdotally, we use a different implementation of the `DTLSListener` internally so introduce additional functionality). This PR maintains the existing `go-coap` functionality, while also demonstrating how a consumer can provide their own listener for custom functionality (including connection IDs).

It seems like previously the `pion/dtls.Listener` was [used](https://github.com/plgd-dev/go-coap/pull/379/files#diff-aaa722ddf89aaefa433bb78b18f5e8f69544665651b12bfa78d1e6dc57a2f1a6L42), but then was moved away from in order to introduce parallel handshakes. My understanding of this functionality is that the `DTLSListener` essentially just continuously accepts connections so that the handshake can begin and complete, perhaps before the DTLS `Server` calls `AcceptWithContext()`. It seems the major blocker from using the `pion/dtls.Listener` is the [starting of the `pion/dtls.Server` [in a go pool](https://github.com/plgd-dev/go-coap/blob/12f9eb2affdaa01ec1a7c64cef4b3dad41f39241/net/dtlslistener.go#L123). My question is whether this could be functionality introduced in `pion/dtls` instead (i.e. stop blocking `Accept()` on completion of the handshake), which would allow `go-coap` to treat the DTLS listener as just a `net.Listener` (with the generic wrapping to support context handling).

Reply:
(Note: this is what I would like to do long term, but the changes in this PR enable folks to implement custom behavior to enable connection ID support, while keeping the previous behavior of `go-coap` unchanged when using the provided `DTLSListener` implementation. This feels like a good intermediate step to me as it does not introduce any breaking changes in `go-coap`, but I am open to discussion and happy to explore other avenues as well!)

Reply:
> It seems like previously the pion/dtls.Listener was [used](https://github.com/plgd-dev/go-coap/pull/379/files#diff-aaa722ddf89aaefa433bb78b18f5e8f69544665651b12bfa78d1e6dc57a2f1a6L42), but then was moved away from in order to introduce parallel handshakes. My understanding of this functionality is that the DTLSListener essentially just continuously accepts connections so that the handshake can begin and complete, perhaps before the DTLS Server calls AcceptWithContext(). It seems the major blocker from using the pion/dtls.Listener is the [starting of the pion/dtls.Server [in a go pool](https://github.com/plgd-dev/go-coap/blob/12f9eb2affdaa01ec1a7c64cef4b3dad41f39241/net/dtlslistener.go#L123). My question is whether this could be functionality introduced in pion/dtls instead (i.e. stop blocking Accept() on completion of the handshake), which would allow go-coap to treat the DTLS listener as just a net.Listener (with the generic wrapping to support context handling).

There is a discussion (https://github.com/pion/dtls/issues/279) about moving the handshake, as it is in the `golang/crypto` package, to the `Read/Write` functions.

I agree with you; we can proceed step by step.